db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java

```diff
@@ -26,348 +26,400 @@
+        failure(task.get().getFailureHandler(), execution, unhandledError, executionStarted);
+        statsRegistry.register(StatsRegistry.ExecutionStatsEvent.FAILED);
+      }
+    }
 
-	}
+    private void complete(CompletionHandler completion, Execution execution, Instant executionStarted) {
+      ExecutionComplete completeEvent = ExecutionComplete.success(execution, executionStarted, clock.now());
+      try {
+        completion.complete(completeEvent, new ExecutionOperations(taskRepository, execution));
+        statsRegistry.registerSingleCompletedExecution(completeEvent);
+      } catch (Throwable e) {
+        statsRegistry.register(SchedulerStatsEvent.COMPLETIONHANDLER_ERROR);
+        statsRegistry.register(SchedulerStatsEvent.UNEXPECTED_ERROR);
+        LOG.error("Failed while completing execution {}. Execution will likely remain scheduled and locked/picked. " +
+            "The execution should be detected as dead in {}, and handled according to the tasks DeadExecutionHandler.", execution, getMaxAgeBeforeConsideredDead(), e);
+      }
+    }
 
-	public static SchedulerBuilder create(DataSource dataSource, Task<?> ... knownTasks) {
-		return create(dataSource, Arrays.asList(knownTasks));
-	}
+    private void failure(FailureHandler failureHandler, Execution execution, Throwable cause, Instant executionStarted) {
+      ExecutionComplete completeEvent = ExecutionComplete.failure(execution, executionStarted, clock.now(), cause);
+      try {
+        failureHandler.onFailure(completeEvent, new ExecutionOperations(taskRepository, execution));
+        statsRegistry.registerSingleCompletedExecution(completeEvent);
+      } catch (Throwable e) {
+        statsRegistry.register(SchedulerStatsEvent.FAILUREHANDLER_ERROR);
+        statsRegistry.register(SchedulerStatsEvent.UNEXPECTED_ERROR);
+        LOG.error("Failed while completing execution {}. Execution will likely remain scheduled and locked/picked. " +
+            "The execution should be detected as dead in {}, and handled according to the tasks DeadExecutionHandler.", execution, getMaxAgeBeforeConsideredDead(), e);
+      }
+    }
 
-	public static SchedulerBuilder create(DataSource dataSource, List<Task<?>> knownTasks) {
-		return new SchedulerBuilder(dataSource, knownTasks);
-	}
+  }
+
+  public static SchedulerBuilder create(DataSource dataSource, Task<?>... knownTasks) {
+    return create(dataSource, Arrays.asList(knownTasks));
+  }
+
+  public static SchedulerBuilder create(DataSource dataSource, List<Task<?>> knownTasks) {
+    return new SchedulerBuilder(dataSource, knownTasks);
+  }
+
+  private static ThreadPoolExecutor newThreadPool(String prefix, int maxThreads, Duration threadTimeout) {
+    ThreadPoolExecutor executor = new ThreadPoolExecutor(maxThreads, maxThreads,
+        threadTimeout.toMillis(), TimeUnit.MILLISECONDS,
+        new LinkedBlockingQueue<>(),
+        defaultThreadFactoryWithPrefix(prefix)
+    );
+    executor.allowCoreThreadTimeOut(true);
```

Comment:
Currently, I don't think that the pool will ever be scaled to less than 3 threads? (Since currently all util-threads run until shutdown with their own sleep)